### PR TITLE
Upgrade Compiling/Packaging to Typescript/Webpack

### PR DESCRIPTION
**Pull request by hawkinsw**
1. Slowly begin migrating code to Typescript from raw Javascript
2. Instead of relying on built-in node module bundling, use webpack
3. Add a `yarn watch` command that will rebuild mw2pdf when any TS
   file is edited. (see https://github.com/OpenTechStrategies/torque-devenv/issues/19)

These are prerequisites (in a very yak-shaving way) for adding unit
testing to the code via Jest. See https://github.com/OpenTechStrategies/SimpleBook/issues/20.

### CODE REVIEW

File: services/api/mw2pdf/src/classes/MediaWikiSession.ts [hunk 1]
```diff
@@ -1,22 +1,25 @@
 import puppeteer from 'puppeteer'
 import {
   PdfGenerator
-} from './PdfGenerator.js'
```
Comment (slifty) [inline services/api/mw2pdf/src/classes/MediaWikiSession.ts:4]: I'm so happy to see `.js` lines go away.

File: services/api/mw2pdf/package.json [hunk 2]
```diff
@@ -2,11 +2,11 @@
   "name": "mw2pdf",
   "version": "0.0.1",
   "description": "Given mediawiki URL, output PDF",
-  "main": "index.js",
+  "private": true,
```
Comment (slifty) [inline services/api/mw2pdf/package.json:5]: We (or maybe just I) do hope to eventually publish mw2pdf as a standalone package.

We haven't done that yet, so I'm fine with this line though I admit I don't know what it gets us aside from protection from accidental publishing.

File: services/api/mw2pdf/INSTALL.md [hunk 3]
```diff
@@ -1,8 +1,12 @@
...
-## Running the script
+# Building
 
-You can run the script using `node index.js`
\ No newline at end of file
+`yarn build`
+
+# Running
+
+You can run the script using `node built/main.js`
```
Comment (slifty) [inline services/api/mw2pdf/INSTALL.md]: While we're changing the entry point to the script: lets add a yarn script to run this command (in case it ever changes again), and then update the python server to call `yarn start` or whatever instead of `node index.js`.

This should be a separate commit IMO but should be part of this PR.

File: services/api/mw2pdf/INSTALL.md [hunk 4]
```diff
@@ -1,8 +1,12 @@
...
 This script requires node.js to be installed.
-In order to run the script, you will first need to build the dependencies using `npm install`
+In order to run the script, you will first need to build the dependencies using `yarn install`
 
-## Running the script
+# Building
 
-You can run the script using `node index.js`
\ No newline at end of file
+`yarn build`
```
Comment (slifty) [inline services/api/mw2pdf/INSTALL.md:8]: Not for this PR, but maybe a good immediate followup task: now that building is a thing... we should probably add something to devenv that automatically builds this package whenever files are changed.

Could you make an issue in devenv to that effect?

File: services/api/mw2pdf/package.json [hunk 5]
```diff
@@ -19,5 +19,12 @@
     "pdfmake": "^0.1.68",
     "puppeteer": "^5.5.0",
     "uuid": "^8.3.2"
+  },
+  "devDependencies": {
+    "awesome-typescript-loader": "^5.2.1",
```
Comment (slifty) [inline services/api/mw2pdf/package.json]: I don't know webpack too well but wanted to sanity check: it looks like `awesome-typescript-loader` is a fork of `ts-loader` with about 10% of the user base (300k vs 3.2m).

More importantly: the awesome variety's last push is 3 years ago whereas ts-loader was 9 days ago.

If you actively picked this one understood and I trust your judgement!  If it was just a "this was the one that I've used in the past" then it might be worth considering using [ts-loader](https://www.npmjs.com/package/ts-loader) instead.
I just wanted to make sure we want to use that fork for whatever reason.

**Review by slifty — CHANGES_REQUESTED** [review]
Yes!!!

Thank you for this work, it's excellent and moving to typescript is a great step forward for the project.

Some small comments / questions but the only meaningful action items are:

1) Considering the use of `ts-loader` since it seems to be more supported.
2) Making sure the new entrypoint doesn't break the python call (check `app.py`).
3) Spinning up an issue in devenv to have rebuilds-on-file-changes be automatic.

File: services/api/mw2pdf/package.json (hunk 2 above)
Comment (hawkinsw) [reply services/api/mw2pdf/package.json:5]: I agree that we do ultimately want to publish as a package, but I think that prevention from accidental publishing is *exactly* what it gets you -- a Good Thing (tm) because I am error prone. If you want me to remove this line, that is perfectly okay. Just let me know your preference. :-)

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md]: Great. What do you think about `yarn run` or something like that?

File: services/api/mw2pdf/INSTALL.md (hunk 4 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md:8]: I'm sorry but I don't quite understand what you are asking here.

File: services/api/mw2pdf/package.json (hunk 5 above)
Comment (hawkinsw) [reply services/api/mw2pdf/package.json]: I think that I decided to use it for exactly 1 reason -- that I used it in a previous project and I was cargo culting. Let's go with ts-loader at your suggestion. I agree that going with the mainstream is a good idea! Thanks for doing this research and checking me on it!

File: services/api/mw2pdf/package.json (hunk 2 above)
Comment (slifty) [reply services/api/mw2pdf/package.json:5]: Let's keep it!  We can remove it if / when we're ready to publish.

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (slifty) [reply services/api/mw2pdf/INSTALL.md]: `yarn run` sounds perfect!

File: services/api/mw2pdf/INSTALL.md (hunk 4 above)
Comment (slifty) [reply services/api/mw2pdf/INSTALL.md:8]: Sorry about that!

So basically in the current setup we have:

1) `SimpleBook` mounted into devenv
2) Python triggering a node process that runs `node index.js {book params}`

This means in the current setup if I were to edit `index.js` that change would actually be reflected on the development environment processes without me taking any extra steps.  If I triggered a new book, that new book would be processed using whatever logic I have.

Now that we're doing typescript / webpack I'm under the impression that flow won't be quite as seemless (by default).  Specifically: the typescript has to be built into javascript after each change in order for the change to be reflected in the live process.

I'm just suggesting we should:

1) add something like `nodemon` or whatever, which will watch the source directory and trigger `yarn build` any time a source file is changed.  This could be wrapped in a `yarn watch` command, or not.
2) add a `supervisor` task to devenv which keeps that watch-and-build process running.

The goal is that we don't want developers to have to manually run `yarn build` every time they make a change and want to test it locally.

Hopefully this makes more sense but let me know if not.

File: services/api/mw2pdf/src/classes/MediaWikiSession.ts (hunk 1 above)
Comment (hawkinsw) [reply services/api/mw2pdf/src/classes/MediaWikiSession.ts:4]: Me, too!!

File: services/api/mw2pdf/package.json (hunk 2 above)
Comment (hawkinsw) [reply services/api/mw2pdf/package.json:5]: Sounds good!

File: services/api/mw2pdf/package.json (hunk 5 above)
Comment (hawkinsw) [reply services/api/mw2pdf/package.json]: Resolved in an updated version of the PR. Again, thanks!!

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md]: Turns out that `run` is a subcommand of `yarn` that cannot be "hijacked" by naming a script that in `package.json`. The same goes for `exec`. The next best thing I could think of was `main`, so that the user would run `yarn main`. What do you think about this?

**Comment by hawkinsw** [comment]
> Yes!!!
> 
> Thank you for this work, it's excellent and moving to typescript is a great step forward for the project.

Thank you for the feedback!

> 
> Some small comments / questions but the only meaningful action items are:
> 
>     1. Considering the use of `ts-loader` since it seems to be more supported.

Done!

> 
>     2. Making sure the new entrypoint doesn't break the python call (check `app.py`).

I _think_ that we are going to need to have a separate PR to update the `app.py` when this is merged. Are you opposed to that?

> 
>     3. Spinning up an issue in devenv to have rebuilds-on-file-changes be automatic.

Will work this with you separately. 

Thanks again for the review!
Will

File: services/api/mw2pdf/INSTALL.md (hunk 4 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md:8]: > Sorry about that!

Not your fault!

> 
> So basically in the current setup we have:
> 
>     1. `SimpleBook` mounted into devenv
> 
>     2. Python triggering a node process that runs `node index.js {book params}`
> 
> 
> This means in the current setup if I were to edit `index.js` that change would actually be reflected on the development environment processes without me taking any extra steps. If I triggered a new book, that new book would be processed using whatever logic I have.
> 
> Now that we're doing typescript / webpack I'm under the impression that flow won't be quite as seemless (by default). Specifically: the typescript has to be built into javascript after each change in order for the change to be reflected in the live process.

You are absolutely correct!

> 
> I'm just suggesting we should:
> 
>     1. add something like `nodemon` or whatever, which will watch the source directory and trigger `yarn build` any time a source file is changed.  This could be wrapped in a `yarn watch` command, or not.
> 
>     2. add a `supervisor` task to devenv which keeps that watch-and-build process running.
> 
> 

This sounds like a good plan!

> The goal is that we don't want developers to have to manually run `yarn build` every time they make a change and want to test it locally.

Agreed.

> 
> Hopefully this makes more sense but let me know if not.

This makes much more sense! 

I will open an issue in devenv to do this work!

Will

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md:12]: Turns out that `run` is a subcommand of `yarn` that cannot be "hijacked" by naming a script that in `package.json`. The same goes for `exec`. The next best thing I could think of was `main`, so that the user would run `yarn main`. What do you think about this?

File: services/api/mw2pdf/INSTALL.md (hunk 4 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md:8]: This is now done via https://github.com/OpenTechStrategies/torque-devenv/issues/19. I will let @slifty resolve this when they think that it is resolved.

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (slifty) [reply services/api/mw2pdf/INSTALL.md]: `yarn main` feels more python than node somehow / at that point I'd just stick with `yarn start` which also feels awkward but is more standard -- but that's my bikeshed-painting-brain!

`yarn main` totally works!

---

Another thought though: what about something more functionality-specific like `yarn mw2pdf` or `yarn pdf`?

Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md]: I have changed it so that the command is `yarn start` with aliases to it for `yarn mw2pdf` and `yarn pdf`. Let me know what you think, @slifty .

**Comment by hawkinsw** [comment]
> > Yes!!!
> > Thank you for this work, it's excellent and moving to typescript is a great step forward for the project.
> 
> Thank you for the feedback!
> 
> > Some small comments / questions but the only meaningful action items are:
> > ```
> > 1. Considering the use of `ts-loader` since it seems to be more supported.
> > ```
> 
> Done!
> 
> > ```
> > 2. Making sure the new entrypoint doesn't break the python call (check `app.py`).
> > ```
> 
> I _think_ that we are going to need to have a separate PR to update the `app.py` when this is merged. Are you opposed to that?
> 

Let me know what you think here, @slifty. 

Thanks!
Will


> > ```
> > 3. Spinning up an issue in devenv to have rebuilds-on-file-changes be automatic.
> > ```
> 
> Will work this with you separately.
> 
> Thanks again for the review!
> Will

**Comment by slifty** [comment]
@hawkinsw The only change I would still ask for is related to:

> 2. Making sure the new entry point doesn't break the python call (check `app.py`).

You mentioned this might need to be a separate PR but if we do that then this PR would result in a broken build (temporarily, but still).

To be clear what I'm asking: basically that this block be changed: https://github.com/OpenTechStrategies/SimpleBook/blob/99bfcf66e30db9053eb675aac84830a9ba23920a/services/api/app.py#L20-L22

to something like:

```
    params = [
        "yarn",
        "--cwd", "mw2pdf"
        "main",
        ...
    ]
```

^ note that [--cwd](https://malki.dev/run-yarn-command-in-a-different-directory-and-starting-up-2-servers-with-node-yarn/) lets you run a yarn command from a different directory.

(where `main` is whatever you end up using to be the equivalent of `start`)

**that said** if I'm missing something please feel free to ignore this request and do it as a separate PR!  This is lookin' good.

**Comment by hawkinsw** [comment]
> @hawkinsw The only change I would still ask for is related to:
> 
> > 1. Making sure the new entry point doesn't break the python call (check `app.py`).
> 
> You mentioned this might need to be a separate PR but if we do that then this PR would result in a broken build (temporarily, but still).
> 
> To be clear what I'm asking: basically that this block be changed:
> 
> https://github.com/OpenTechStrategies/SimpleBook/blob/99bfcf66e30db9053eb675aac84830a9ba23920a/services/api/app.py#L20-L22
> 
> to something like:
> 
> ```
>     params = [
>         "yarn",
>         "--cwd", "mw2pdf"
>         "main",
>         ...
>     ]
> ```
> 
> ^ note that [--cwd](https://malki.dev/run-yarn-command-in-a-different-directory-and-starting-up-2-servers-with-node-yarn/) lets you run a yarn command from a different directory.
> 
> (where `main` is whatever you end up using to be the equivalent of `start`)
> 
> **that said** if I'm missing something please feel free to ignore this request and do it as a separate PR! This is lookin' good.

After our interactive conversation earlier today I now realize what you meant. I have made the change in the most recent version of this PR. Let me know what you think!

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (slifty) [reply services/api/mw2pdf/INSTALL.md]: Nice!

Comment (slifty) [inline services/api/mw2pdf/INSTALL.md]: I think we can  / should update this to `yarn start` now that we have that in our yarn.

**Comment by slifty** [comment]
Looks great!  May wanna update the documented run command but fine either way!

Regardless no need for re-review; merge away when ready (though of course there are now branch conflicts :()

**Review by slifty — APPROVED** [review]

File: services/api/mw2pdf/INSTALL.md (hunk 3 above)
Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md]: Resolving given :rocket:

Comment (hawkinsw) [reply services/api/mw2pdf/INSTALL.md]: Done!!

**Comment by slifty** [comment]
Just a note: I am in the process of revamping SimpleBook's installation in dev-env to use Ansible, and so we're waiting on that to get pushed before we can do the final test-and-merge of this PR.  Stay tuned!

**Comment by hawkinsw** [comment]
> Just a note: I am in the process of revamping SimpleBook's installation in dev-env to use Ansible, and so we're waiting on that to get pushed before we can do the final test-and-merge of this PR. Stay tuned!

And, as I said in "chat", take your time! I want to make sure that everything works so that I don't screw things up!

Will
